### PR TITLE
added parse_resource functionality as a utility function

### LINKED ISSUE #577: Expose a parse_resource utility function for plugins.
In [kinto-signer](https://github.com/Kinto/kinto-signer/blob/25a3979/kinto_signer/utils.py#L14-L24) and [kinto-amo](https://github.com/mozilla-services/kinto-amo/blob/a1801b2efee0d95b41f9b8ffe285148077fc3864/kinto_amo/utils.py#L5-L14) we are using ``parse_resource`` to extract from a configuration the bucket and collection of a given resource.

Since two plugins start to using the same thing, it is probably time to make it available as a plugin utility in Kinto itself.

### PR DESCRIPTION
**Pull request by oronsan**
Fixes #577 

as per your guidelines, I submitted a pull request, but my contribution is not yet complete
I'll work on all the requirements later today

- [x] Add documentation.
- [x] Add tests.
- [x] Add a changelog entry.
- [x] Add your name in the contributors file.
- [x] If you changed the HTTP API, update the [API_VERSION](https://github.com/Kinto/kinto/blob/master/kinto/__init__.py#L15) constant and add an API changelog entry [in the docs](https://github.com/Kinto/kinto/blob/master/docs/api/index.rst)

r? @

…urce function

### CODE REVIEW

**Comment by leplatrem** [comment]
Thanks for starting this!

I think you could put it in `kinto/core/utils.py`. It should also have some tests to make sure it behaves as expected.

Don't hesitate to ask specific questions here if you're stuck!

**Comment by oronsan** [comment]
thanks, @leplatrem 
I appreciate that :)

I saw that util.py file you mentioned, and wasn't sure if it fits there

**Comment by oronsan** [comment]
I am done working on the feature, according to the guidelines.
Please review my work @leplatrem  

thank you :)

File: CONTRIBUTORS.rst [hunk 1]
```diff
@@ -53,3 +53,4 @@ Contributors
 * Varna Suresh <varna96@gmail.com>
 * Wil Clouser <wclouser@mozilla.com>
 * Yann Klis <yann.klis@gmail.com>
+* Oron Gola <oron.gola@gmail.com>
```
Comment (Natim) [inline CONTRIBUTORS.rst]: Could you put your name in an alphabetic order in the list please?

Comment (oronsan) [inline CONTRIBUTORS.rst]: done, please review

File: tests/core/test_utils.py [hunk 2]
```diff
@@ -246,3 +247,21 @@ def test_ignores_if_subobject_is_not_dict(self):
...
+    def test_parse_resource(self):
+        bad_inputs = ['none', '/', '/one', 'one/', '/tw/o']
+        good_inputs = ['bid/cid', '/buckets/bid/collections/cid']
+        expected = {
+            'bucket': 'bid',
+            'collection': 'cid'
+        }
+        for input in bad_inputs:
+            with pytest.raises(ValueError):
+                parse_resource(input)
```
Comment (oronsan) [inline tests/core/test_utils.py]: done, please review

File: tests/core/test_utils.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 from __future__ import unicode_literals
 import unittest
 import os
+import pytest
```
Comment (gabisurita) [inline tests/core/test_utils.py]: Just a small tip: you shouldn't include pytest here.

You can run tests using:
- ``pytest tests/path/to/my/test`` under the virtualenv to run a single test. 
- ``pytest tests/`` under the virtualenv to run all the tests.
- ``make tests`` to run all the tests for all python versions, like travis do.

Hope this helps! :)

Comment (oronsan) [inline tests/core/test_utils.py]: hi,

thanks for the tip.
I didn't understand something:

I did use make tests to run all the test in the project.
but how could I avoid importing pytest?

I have used it in the same file on line 266:
      with pytest.raises(ValueError) as excinfo:

so wasn't I supposed to import pytest?
is there another way to reference it?

Thank you!

Comment (oronsan) [inline tests/core/test_utils.py]: Ok, I got it.
Found a different solution with asserRaises()

Thank you for the tip :)

Comment (gabisurita) [inline tests/core/test_utils.py]: Oh, I'm really sorry. I didn't get that. :/

But why is pytest.raises better than assertRaises?

Comment (oronsan) [inline tests/core/test_utils.py]: Ok, so I switched back to pytest.raises and got it working now :)

File: docs/core/utils.rst [hunk 4]
```diff
@@ -4,6 +4,8 @@
 Utils
 #####
 
+The ``parse_resource`` function provides a way to break a resource 
+into its bucket and collection id parts
```
Comment (leplatrem) [inline docs/core/utils.rst]: Why the ``parse_resource()`` would be documented differently? This is not necessary IMO

File: kinto/core/utils.py [hunk 5]
```diff
@@ -423,3 +423,24 @@ def instance_uri(request, resource_name, **params):
     """Return the URI for the given resource."""
     return strip_uri_prefix(request.route_path('%s-record' % resource_name,
                                                **params))
+
+
+def parse_resource(resource):
+    """extract the bucket and collection of a given resource"""
```
Comment (leplatrem) [inline kinto/core/utils.py]: * Nitpick: capital `Extract`
* What is resource here? → It's a URI
* What are bucket and collection → they are bucket id and collection id
* What does the function return? A tuple? A dict ?

File: tests/core/test_utils.py [hunk 6]
```diff
@@ -246,3 +247,26 @@ def test_ignores_if_subobject_is_not_dict(self):
...
+                         "'/buckets/<bid>/collections/<cid>' or "
+                         "'<bid>/<cid>'. Got %r" % input)
+
+            with pytest.raises(ValueError) as excinfo:
+                parse_resource(input)
+            self.assertEquals(str(excinfo.value), error_msg)
+
+        for input in good_inputs:
+            parts = parse_resource(input)
+            self.assertEqual(expected, parts)
```
Comment (leplatrem) [inline tests/core/test_utils.py]: You could just take the tests from here: https://github.com/Kinto/kinto-signer/blob/06f790cc57ad3b53db090d62a5318df96e320fdd/tests/test_utils.py#L9-L130

Comment (oronsan) [inline tests/core/test_utils.py]: Thank you for the review @leplatrem 
I changed the files according to your requests

**Comment by oronsan** [comment]
is it possible that there's a problem with the build process with the docs job?
I didn't do anything in particular to make to break, and it's not passing
I noticed that other recent build jobs have this issue

thank you

**Comment by leplatrem** [comment]
> is it possible that there's a problem with the build process with the docs job?

Yes you are totally right! I can't spend time on this right now unfortunately.

For the tests, why didn't you just copy https://github.com/Kinto/kinto-signer/blob/06f790cc57ad3b53db090d62a5318df96e320fdd/tests/test_utils.py#L9-L130 ?
Since you copied the code of the `parse_resource()` it would make sense to copy the tests too no?

**Comment by oronsan** [comment]
I did copy, but not exactly
the code there assumes the format is source_url;dest_url i.e:
/buckets/sbid/collections/scid;/buckets/dbid/collections/dcid

and my code only expects a single url.
so I only took the relevant tests from there

do you think it's necessary to have source;dest format?

**Comment by leplatrem** [comment]
> do you think it's necessary to have source;dest format?

OH I'm very sorry! I hadn't caught that!

**Pull request merged**